pkg/container/qemu_runner.go

```diff
@@ -0,0 +1,744 @@
+			if err != nil {
+				clog.FromContext(ctx).Errorf("qemu: could not inject needed kernel modules into initramfs: %v", err)
+				return "", err
+			}
+		}
+	}
+
+	err = apko_cpio.FromLayer(layer, guestInitramfs)
+	if err != nil {
+		return ref, fmt.Errorf("failed to create cpio initramfs: %w", err)
+	}
+
+	return guestInitramfs.Name(), nil
+}
+
+func (b qemuOCILoader) RemoveImage(ctx context.Context, ref string) error {
+	clog.FromContext(ctx).Infof("removing image path %s", ref)
+	return os.RemoveAll(ref)
+}
+
+func createMicroVM(ctx context.Context, cfg *Config) error {
+	clog.FromContext(ctx).Debug("qemu: ssh - create ssh key pair")
+	pubKey, err := generateSSHKeys(ctx, cfg)
+	if err != nil {
+		return err
+	}
+
+	// always be sure to create the VM rootfs first!
+	kernelPath, rootfsInitrdPath, err := getKernelPath(ctx, cfg)
+	if err != nil {
+		clog.FromContext(ctx).Errorf("could not prepare rootfs: %v", err)
+		return err
+	}
+
+	baseargs := []string{}
+	// load microvm profile and bios, shave some milliseconds from boot
+	// using this will make a complete boot->initrd (with working network) In ~700ms
+	// instead of ~900ms.
+	if _, err := os.Stat("/usr/share/qemu/bios-microvm.bin"); err == nil {
+		// only enable pcie for network, enable RTC for kernel, disable i8254PIT, i8259PIC and serial port
+		baseargs = append(baseargs, "-machine", "microvm,rtc=on,pcie=on,pit=off,pic=off,isa-serial=off")
+		baseargs = append(baseargs, "-bios", "/usr/share/qemu/bios-microvm.bin")
+	}
+
+	baseargs = append(baseargs, "-kernel", kernelPath)
+	baseargs = append(baseargs, "-initrd", rootfsInitrdPath)
+
+	if cfg.Memory != "" {
+		memKb, err := convertMemoryToKB(cfg.Memory)
+		if err != nil {
+			return err
+		}
+
+		baseargs = append(baseargs, "-m", fmt.Sprintf("%dk", memKb))
+	} else {
+		baseargs = append(baseargs, "-m", getAvailableMemoryKB())
+	}
+
+	if cfg.CPU != "" {
+		baseargs = append(baseargs, "-smp", cfg.CPU)
+	} else {
+		baseargs = append(baseargs, "-smp", fmt.Sprintf("%d", runtime.NumCPU()))
+	}
+
+	// use kvm on linux, and Hypervisor.framework on macOS
+	if runtime.GOOS == "linux" {
+		baseargs = append(baseargs, "-enable-kvm")
+	} else if runtime.GOOS == "darwin" {
+		baseargs = append(baseargs, "-accel", "hvf")
+	}
+
+	baseargs = append(baseargs, "-cpu", "host")
+	baseargs = append(baseargs, "-daemonize")
+	// ensure we disable unneeded devices, this is less needed if we use microvm machines
+	// but still useful otherwise
+	baseargs = append(baseargs, "-display", "none")
+	baseargs = append(baseargs, "-no-reboot")
+	baseargs = append(baseargs, "-no-user-config")
+	baseargs = append(baseargs, "-nodefaults")
+	baseargs = append(baseargs, "-parallel", "none")
+	baseargs = append(baseargs, "-serial", "none")
+	baseargs = append(baseargs, "-vga", "none")
+	// use -netdev + -device instead of -nic, as this is better supported by microvm machine type
+	baseargs = append(baseargs, "-netdev", "user,id=id1,hostfwd=tcp::"+cfg.SSHPort+"-:22")
```

Comment:
For future, and micro optimization.  In order to avoid the "find a random , hopefully  unused port" thing, you can do 
1. pick a random 127.0.0.0/8 address - you can use port 22 on any one of the 16 million (2^24) host addresses on the localhost network. 
2. use port 0 and then ask the qemu monitor which one it got.

Reply:
Implemented this method instead :+1: